sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java

```diff
@@ -239,8 +246,53 @@ private TestPipeline(
   public Statement apply(final Statement statement, final Description description) {
     return new Statement() {
 
+      private final Predicate<Annotation> hasCategoryAnnotation = new Predicate<Annotation>() {
+
+        @Override
+        public boolean apply(final Annotation annotation) {
+          return annotation.annotationType() != null
+              && annotation.annotationType().equals(Category.class);
+        }
+      };
+
+      private final Predicate<Annotation> categoryOfNeedsRunner = new Predicate<Annotation>() {
+
+        @Override
+        public boolean apply(final Annotation category) {
+          return
+              FluentIterable
+                  .from(Arrays.asList(((Category) category).value()))
+                  .anyMatch(new Predicate<Class<?>>() {
+
+                    @Override
+                    public boolean apply(final Class<?> aClass) {
+                      return NeedsRunner.class.isAssignableFrom(aClass);
+                    }
+                  });
+        }
+      };
+
+      private void setDeducedEnforcementLevel() {
+        if (enforcement instanceof PipelineAbandonedNodeEnforcement) {
```

Comment:
Not exactly. `setDeducedEnforcementLevel()` "knows" that the default enforcement is `PipelineAbandonedNodeEnforcement`, so in case it detects it's been called with the enforcement not being the default it will not run the inference logic and leave things as they are, or better yet, as the user has set them up.

```java 
if (enforcement instanceof PipelineAbandonedNodeEnforcement /* which is the default */ ) {
 // do inference 
} else /* enforcement is not the default, so the user must have changed it */ { 
  // bail out, do not execute inference logic
}
```